Prevent removing last item in order

guidoDutra opened the pull request:
#### What? Why?

Closes #7627 

Prevents user from deleting last item from order in admin order edit page. See #7627.

#### What should we test?
When deleting the last item from an order, in both scenarios described in issue #7627, user should see an alert and the operation should not be completed.

#### Release notes

Prevents user from deleting last item from an order in admin order edit page.

Changelog Category: User facing changes

guidoDutra commented on app/assets/javascripts/admin/spree/orders/variant_autocomplete.js.erb:
```diff
@@ -54,46 +54,53 @@ $(document).ready(function() {
+  var inventory_units = _.where(shipment.inventory_units, {variant_id: variant_id});
+
+  if (quantity == 0 && inventory_units.length == shipment.inventory_units.length) {
+    alert(t("js.admin.orders.cannot_remove_last_item"));
```
The issue mentions that "the warning Line items can't be blank should apply", which refers to a Rails flash error message (see Issue). However, I think it's easier and safer to validate this in the frontend (avoids interfering with other parts of the site), so I used an alert, as it being done elsewhere in this page. Is this ok?

andrewpbrett replied:
That seems fine to me - it's more descriptive this way as well 👍

jibees replied:
Does this `alert` looks like the default javascript alert modal?
Actually I don't know if it used somewhere else, and I'm not a huge fan of this default modal. 
I can see in the code a `info_dialog` maybe it's a good way to do this?

andrewpbrett commented on app/assets/javascripts/admin/spree/orders/variant_autocomplete.js.erb:
```diff
@@ -54,46 +54,53 @@ $(document).ready(function() {
+  var url = Spree.routes.orders_api + "/" + order_number + "/shipments/" + shipment_number;
+
+  var new_quantity = 0;
+  if(inventory_units.length<quantity){
+    url += "/add"
+    new_quantity = (quantity - inventory_units.length);
+  }else if(inventory_units.length>quantity){
```
Very minor style suggestions:

```suggestion
  if (inventory_units.length < quantity) {
    url += "/add";
    new_quantity = (quantity - inventory_units.length);
  } else if (inventory_units.length > quantity) {
```

guidoDutra replied:
thanks @andrewpbrett, just squashed this in

andrewpbrett commented on app/assets/javascripts/admin/spree/orders/variant_autocomplete.js.erb:
```diff
@@ -54,46 +54,53 @@ $(document).ready(function() {
+      }
     }
-    $('a.delete-item').click(handle_delete_click);
+    $('a.delete-item').on('confirm:complete', handle_delete_click);
```
I trust that this works but I'm curious what's actually happening here - we're now relying on the return value of `adjustItems`, which will be false if we shouldn't remove the last item, is that basically it?

guidoDutra replied:
This is just to prevent a bug: the alert box that shows up when the user clicks the delete button was not having any effect, since just clicking the button already trigered the `handle_delete_click` callback. Now its triggered by `confirm:complete`, i. e. when the user clicks the confirm button in the alert. If the user clicks the cancel button the function won't be called. This is also described in the issue:

By editing an order, /admin/orders/<order_number>/edit it is possible to remove the last item here it's possible both by clicking the trash-bin (**somehow clicking OK seems not to be required**) and by reducing the number of items to zero.

andrewpbrett reviewed (approved):
Thanks @guidoDutra! Couple minor notes but this looks good.

jibees commented:
One thing that is really helpful is to add a little video/gif of the new feature (if feasible). Many thanks!

jibees reviewed (commented):
Just add a comment of the look and feel of the alert modal. Wondering if it's a good idea to use the default one. 
Maybe we could add screenshots/video/gif of the feature, and tell the product-owner if it's ok ?

guidoDutra commented:
Hey @jibees, just added a video to the PR description. Regarding the alert, the default alert is used in other parts of the same file (`variant_autocomplete.js.erb`). You can see for example on the video I just added, when the user clicks the delete button. Maybe that's because this is code inherited from sprint? I'm not sure if it would be straightforward to use angular in these kind of files, but I can investigate if needed.

jibees commented:
> Regarding the alert, the default alert is used in other parts of the same file (variant_autocomplete.js.erb)

If it's already used elsewhere (and specially in this page), I'm happy to say that this is ready to test then!

filipefurtad0 commented:
Hey @guidoDutra ,

Thank you for addressing this one!

I was able to reproduce your detailed video on this scenario as well.

One thing to note is on the choice of the modal. This specific modal can be muted by the user - on a different context - which could lead to an apparent "broken" button. This is shown on the video below: 

![Peek 2021-08-24 11-23](https://user-images.githubusercontent.com/49817236/130600854-533459a6-a9a1-48a9-91f9-b1d1d2eb068d.gif)

I think this is not a major thing and this PR definitively prevents the bug from occurring, and closes the issue as well :muscle: 
Moving to ready yo go - thank you!

jibees commented:
> This specific modal can be muted by the user - on a different context - which could lead to an apparent "broken" button. 

That's one of the reason I'm advocating for replacing default `alert` by a custom modal. I'd vote Ready to go and maybe open a new good first issue?

filipefurtad0 commented:
Totally agree @jibees,
Opening a new issue :+1:

filipefurtad0 commented:
Issue here - https://github.com/openfoodfoundation/openfoodnetwork/issues/8098

RachL commented:
Hello!

Thanks Guido for this PR :). I've noticed that we have forgotten a test case when writing up the issue: deleting an item through Bulk Order Management.

I've tested this case with order R345817671 on FR staging: deleting the last item through BOM still ends up with an empty complete order. Now this is what was happening before: nothing to revert here :) 

I'm going to open a new issue for this case, but I was wondering about the solution to apply on the bulk order management page. Asking the user to cancel the order is a bit cumbersome on this page especially because the user cannot cancel an order from there, they have to do back-and-force with the order edit page.

Could we instead force the status to canceled, with a modal e.g. "deleting this item will cancel the order as it is the last one, do you wish to proceed: yes/no"?

ping order states team @lin-d-hop @filipefurtad0 @jaycmb

jaycmb commented:
> deleting the last item through BOM still ends up with an empty complete order. 

Not sure if I know where exactly you saw this @RachL. As Filipe pointed out above in BOM the user cannot put the quantity to 0 -> input field gets red outline and alert pops up:

<img width="1290" alt="PastedGraphic-1" src="https://user-images.githubusercontent.com/39744592/131858394-062a1409-b31f-4d93-9671-d034b6baff54.png">

Text of alert could be changed to _"Quantity cannot be 0"_ though to make the error easier to understand for the user.

> Could we instead force the status to canceled, with a modal e.g. "deleting this item will cancel the order as it is the last one, do you wish to proceed: yes/no"?

From a usability perspective makes sense to change the **order status** to `cancelled` (and **shipment status** to `none`) and save the extra steps for the user, but if we do so then in my opinion this same logic should also apply to the order page instead of having 2 processes for the same use case

And then the alerts would also be unified, because now we have 
(1) Order Page: modal 
(2) BOM: error bar in the bottom

But that´s certainly more than the original scope :)

RachL commented:
@jaycmb 

>  As Filipe pointed out above in BOM the user cannot put the quantity to 0 -> input field gets red outline and alert pops up:

Indeed, but you can **delete** the item with the little bin on the right side (that's one of the most useful BOM feature for a hub btw - to be able to delete products from orders quickly). I've tested it, it really does leave the order empty with a complete state.

> then in my opinion this same logic should also apply to the order page instead of having 2 processes for the same use case

Yes I think I agree.

> But that´s certainly more than the original scope :)

Of course. I commented here but my idea was not to reopen the issue, but to open new issues instead. I just wanted to align on the scope of those new issues.

filipefurtad0 commented:
Nice catch @RachL! I've checked the BOM page before, but missed the trash-bin icon. Indeed this leads to an empty order as well...

Agree that we should have a unified warning "strategy" for both pages and throughout the app as you mention @jaycmb. However, maybe we're on the safe side here if we can prevent this from happening in the BOM page:

i) changing the quantity to zero is prevented by the error bar in the bottom - which is **current behavior**
ii) the trash-bin icon is hidden for the last items -  **proposal** as expected behavior in a new issue

Happy to open that issue. What do you think?

jaycmb commented:
> Indeed, but you can delete the item with the little bin icon on the right side

Ah, missed that too. I thought that deleted the order :see_no_evil:
